### PR TITLE
Refactor RA TargetTypes

### PR DESCRIPTION
**Pull request by reaperrr**
One of the main reasons why past attempts to refactor and simplify `CreateEffectWarhead` without dropping support for any edge cases failed was that RA sharing TargetTypes between terrain and actors made things a nightmare once, for example, an effect warhead was meant to trigger on `Water` actors, but not `Water` terrain.
`InvalidTargets` would in that case be insufficient, since terrain most of the time has only one TargetType, while listing all individual actor TargetTypes like `Vehicle`, `Ship` etc. was prone to forgetting one or two.

Cleanly separating between terrain and actors in RA's `TargetTypes`, introducing general `GroundActor` and `WaterActor` target types, as well as not using the `GroundActor` target type on naval units and buildings, makes RA's setup fit for avoiding such issues with the upcoming `CreateEffectWarhead` code simplification.

Also added some minor effect polish in a few cases (`DepthCharge`, ship explosions, `Sniper` weapon).

~Depends on #18066.~

### CODE REVIEW

File: mods/ra/rules/defaults.yaml
Comment (pchote) [inline mods/ra/rules/defaults.yaml]: Shouldn't this be `AirborneActor` for consistency?

**Review by pchote — COMMENTED** [review]
The general idea and implementation LGTM, but I haven't gone through the weapons with the fine toothed comb required to spot any cases that you may have missed so i'll hold off on a :+1: until we can get some testing here (this would be a good place for non-technical people to help) to try and reduce the chance of any regressions slipping in.

File: mods/ra/rules/defaults.yaml
Comment (reaperrr) [reply mods/ra/rules/defaults.yaml]: I guess you're right.

**Comment by reaperrr** [comment]
Updated and rebased.

**Review by netnazgul — CHANGES_REQUESTED** [review]
https://github.com/OpenRA/OpenRA/blob/696069b3a42c36c7f4deb6c7d92059e945cd6733/mods/ra/rules/defaults.yaml#L176-L222
Seems that AutoTarget* virtuals require an update to their ValidTargets and InvalidTargets as well (i.e. AutoTargetAir targets Air, which now supposed to be and/or AirborneActor, I guess).

File: mods/ra/weapons/ballistics.yaml
Comment (netnazgul) [inline mods/ra/weapons/ballistics.yaml:22]: `LeaveSmudge` for `Infantry`?..

File: mods/ra/weapons/explosions.yaml
Comment (netnazgul) [inline mods/ra/weapons/explosions.yaml:16]: Again `LeaveSmudge` question - how does it actually work? What's the reason/use for anything not Ground here?

File: mods/ra/weapons/missiles.yaml
Comment (netnazgul) [inline mods/ra/weapons/missiles.yaml:30]: `LeaveSmudge`, with just Infantry and no Vehicle again.

**Review by netnazgul — COMMENTED** [review]
Can confirm, that AutoTarget (attack and assault move) doesn't work for air-to-air (Longbows)

**Comment by pchote** [comment]
Thinking out loud (not for this pr): do we want `[TargetTypeDefinition]` and `[TargetTypeReference]` attributes and a lint rule to catch cases like the missing `AutoTargetPriority` changes, or would this create too much frustration from false positives like modders often find with conditions?

File: mods/ra/weapons/ballistics.yaml
Comment (reaperrr) [reply mods/ra/weapons/ballistics.yaml:22]: The idea was that infantry shouldn't block explosions from creating smudges, while anything larger should.
Note that it already works like that on bleed, see how the old `InvalidTargets` didn't list `Infantry`.

File: mods/ra/weapons/explosions.yaml
Comment (reaperrr) [reply mods/ra/weapons/explosions.yaml:16]: To make smudges appear only on empty terrain and below the specified actors. These explosions are for actors blowing up, so `Vehicle` has to be valid as well, otherwise vehicle explosions wouldn't leave/grow craters.

File: mods/ra/weapons/missiles.yaml
Comment (reaperrr) [reply mods/ra/weapons/missiles.yaml:30]: See above.

**Comment by reaperrr** [comment]
Updated, fixed auto-targeting of airborne actors.

**Comment by reaperrr** [comment]
> Thinking out loud (not for this pr): do we want [TargetTypeDefinition] and [TargetTypeReference] attributes and a lint rule to catch cases like the missing AutoTargetPriority changes, or would this create too much frustration from false positives like modders often find with conditions?

Hard to say. Something that catches such errors would be nice I guess, at least for `AutoTargetPriority`. Not so sure about weapons and warheads, though.
Would potentially be less painful if we get rid of internal defaults and `[FieldLoader.Require]` all `ValidTargets` fields across the code.

**Review by Smittytron — CHANGES_REQUESTED** [review]
Atom bomb does no damage.

Note to self to update #18098 after this is merged.

**Comment by reaperrr** [comment]
Fixed nukes, looks like I meant to adapt them last and then forgot about them altogether...

**Review by Smittytron — DISMISSED** [review]
LGTM, Didn't spot anything else.

**Comment by Mailaender** [comment]
Torpedos don't show an explosion animations when hitting bridges.

**Comment by Mailaender** [comment]
Tanks don't deal damage against bridges when force firing upon them.

**Comment by reaperrr** [comment]
Fixed bridge-related issues.

**Comment by reaperrr** [comment]
I tested the above fixes myself, and the required yaml changes were minor.
@pchote said on IRC his +1 basically still stands as far as the implementation as such is concerned, and I assume we've caught everything noticable.
If we really overlooked anything, it's likely easy to fix and should turn up soon (or at the latest during the next playtest phase), so I'll go ahead and merge it now.

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e517748fa793c5d1e6ad159f4d0a06da2109c5ee...3eb5b0d7d12ec571d81c0aaf2148169571210420

**Comment by Smittytron** [comment]
Whoops, barrels can't be targeted (creeps?) and don't damage each other.

**Comment by netnazgul** [comment]
@reaperrr ^

**Comment by reaperrr** [comment]
I'll look into it.

**Comment by reaperrr** [comment]
I found out why they don't damage each other (or other stuff, for that matter), but I can't reproduce the "can't be targeted" part, at least in soviet01?

**Comment by Smittytron** [comment]
OK, I was playing on a map that had barrels as Neutral instead of Creeps.

**Comment by Mailaender** [comment]
Does https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/BotModules/Squads/States/AirStates.cs#L21 need an update now?

**Comment by reaperrr** [comment]
Good catch.

Yes, but since only RA uses `AirborneActor` now, it would have to be by unhardcoding and exposing it to yaml.

**Pull request merged**
